### DISCUSSION TITLE
PSD to JPEG conversion ignoring alpha channel while resizing

### DISCUSSION THREAD

**Discussion by jetpack1116** [ImageMagick/ImageMagick#4159] [Help] [answered]
I used the following parameter to ignore the alpha channel of PSD image when converting PSD to JPEG.
```
-define psd:alpha-unblend=off
```
This works OK when I am not resizing the converted JPEG file
```
convert -define psd:alpha-unblend=off c:\testPSDimage.psd[0] -auto-orient -colorspace srgb -background white c:\testPSDimage.jpg
```

But when I resize, `-define psd:alpha-unblend=off` does not work and the output image is the alpha channel.

```
convert -define psd:alpha-unblend=off c:\testPSDimage2.psd[0] -auto-orient -colorspace srgb -background white -resize 1920x1920 c:\testPSDimage2.jpg
```

Has anybody encounter this problem?

Is this an ImageMagick bug? 
Because`-define psd:alpha-unblend=off`  is supposed to be ignoring the alpha channel right? Whether you are resizing the output or not.

**Comment by snibgo** [comment]
> Because-define psd:alpha-unblend=off is supposed to be ignoring the alpha channel right? 

No. On "-define psd:alpha-unblend=off", see the documentation at https://imagemagick.org/script/defines.php . This does not disable the alpha channel. If you want that, then use "-alpha off".

Your "-colorspace srgb" does nothing. Your input PSD has an embedded ICC profile, Adobe1998. If you want sRGB, you should convert it with "-profile sRGB.icc", or some other sRGB ICC file. 

Your "-background white" does nothing.

**Reply by jetpack1116** [reply]
@snibgo  Thank you for your reply.

My goal is to convert PSD image with alpha channels to JPEG image.
The problem is when I use the `-alpha`  options (remove, off, deactivate, disassociate), the resulting image either shows the alpha channel image part only or the whole image but with the alpha channel being traced with a line.

So I use  `-define psd:alpha-unblend=off`.
Using `-define psd:alpha-unblend=off`, the result is the converted JPEG image. (This is what I expected and I got it)
But when I resize, meaning I use  `-define psd:alpha-unblend=off` and `-resize` option, the result was wrong again. I got only the image showing the alpha channel with black background.

I already corrected the other parameters `-colorprofile` `-profile`, `-background` as you mentioned, but still I got the wrong output.

**Comment by fmw42** [comment]
try

convert -define psd:alpha-unblend=off c:\testPSDimage.psd[0] -auto-orient -colorspace srgb -background white -flatten -resize 1920x1920 c:\testPSDimage.jpg`

or

`convert -define psd:alpha-unblend=off c:\testPSDimage.psd[0] -auto-orient -colorspace srgb -background white MIFF:- | convert - -resize 1920x1920 c:\testPSDimage.jpg`

or

`convert -define psd:alpha-unblend=off c:\testPSDimage.psd[0] -auto-orient -colorspace srgb -background white JPG:- | convert - -resize 1920x1920 c:\testPSDimage.jpg`

If that does not work, can you post an example that fails?

**Reply by jetpack1116** [reply]
@fmw42 

sample cmd1: The result shows the alpha channel image only with white background.
sample cmd2: The result shows the alpha channel image only with black background.
sample cmd3: The result shows the converted JPEG image already resized! (This is the expected output)

Wow. Thank you very very much! 👍

**Comment by snibgo** [comment]
> I got only the image showing the alpha channel with black background.

JPEG cannot record transparency. What background do you want? Perhaps white. If so, then flatten against a white blackground before writing the output.

I suggest:
```
magick testPSDimage.psd -profile sRGB.icc -resize 1920x1920 -background White -layers Flatten x2.jpg
```
You can flatten before or after the resize; there is no difference.

Include "-define psd:alpha-unblend=off" if you want.

**Reply by jetpack1116** [reply]
@snibgo this works if I want the alpha channel to display with white background.
Thanks for the explanation, I understand now.

But what I want is just to convert the PSD image to JPEG.
@fmw42  suggestion `convert -define psd:alpha-unblend=off c:\testPSDimage.psd[0] -auto-orient -colorspace srgb -background white JPG:- | convert - -resize 1920x1920 c:\testPSDimage.jpg` works OK now.

Thank you very much for your help.  👍 💯

**Comment by fmw42** [comment]
I do not understand. JPEG cannot have transparency. So what is wrong with the white background?  What background do you expect or want?

**Reply by jetpack1116** [reply]
@fmw42 
We can ignore the background part, its irrelevant to the problem.
My concern is only PSD with alpha channel to JPEG conversion showing only the alpha channel part.
My expectation is the conversion should ignore the alpha channel and output the converted JPEG showing the whole image itself and not the alpha channel only or a trace of the alpha channel.

Your suggestion works for me now.
`convert -define psd:alpha-unblend=off c:\testPSDimage.psd[0] -auto-orient -colorspace srgb -background white JPG:- | convert - -resize 1920x1920 c:\testPSDimage.jpg`

I can even remove `-background white`, its no problem at all.

**Comment by fmw42** [comment]
All that command does is save a write.  It is the same as:

```
convert -define psd:alpha-unblend=off c:\testPSDimage.psd[0] -auto-orient -colorspace srgb -background white temp.jpg 
convert temp.jpg -resize 1920x1920 c:\testPSDimage.jpg

```
then remove temp.jpg

I suggested it as a work around. I do not know why ImageMagick is failing to produce the JPG result directly. That seems like a bug, since JPG cannot have transparency.

Can you post an example PSD file that has this issue? It would help the developers to see what might be the issue.

**Reply by jetpack1116** [reply]
Yup, its quite weird it cannot convert directly to JPEG.


Thanks. :)

**Comment by fmw42** [comment]
ImageMagick cannot use either alpha channel in the file since they are both disabled and if you enable them and close the file, they will get disabled. It can only use background transparency. There is no path either. 

So there is no difference in doing:

`convert -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -background white temp.jpg `

or without the define

`convert testPSDimage.psd -auto-orient -colorspace srgb -background white temp2.jpg `

`compare -metric rmse temp.jpg temp2.jpg null:`
0 (0)

There is only one layer, so no [0] is needed thought would not hurt.

**Reply by fmw42** [reply]
Also no difference when resizing.

```
convert -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -resize 1920x1920 temp.jpg 
convert testPSDimage.psd -auto-orient -colorspace srgb -resize 1920x1920  temp2.jpg 

compare -metric rmse temp.jpg temp2.jpg null:

```
0 (0)

I am using IM 6.9.12-21.  What version are you using?

**Reply by fmw42** [reply]
OK. Must be a new feature in IM 7.  ( 7.1.0-6). As it does use the alpha in that case when resizing only, strangely! And the first version is much better around the border of the masking. This must be either a bug/unwanted behavior or a feature under development. **One of the IM developers will need to comment further** as it was my understanding from talks with the developers that the alpha channels could not be used as they are disabled.

```
magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -resize 1920x1920 temp.jpg 
magick testPSDimage.psd -auto-orient -colorspace srgb -resize 1920x1920  temp2.jpg 

magick compare -metric rmse temp.jpg temp2.jpg null:

```

1071.11 (0.0163441)

**Comment by fmw42** [comment]
You can use PFM format or PNG in place of MIFF. This works for me. PFM is NetPBM format which should be available on most computers. TIFF is not the right format for what you want to do.

`magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -resize 1920x1920 PFM:- | magick - temp.jpg `

Also this works

`magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -resize 1920x1920 PNG:- | magick - temp.jpg `

These all do the masking just like using JPG:-.

**Reply by jetpack1116** [reply]
@fmw42 
Thanks, noted on these suggestions. 👍

**Comment by fmw42** [comment]
I note that convert (IM 6) does not mask using the alpha channel when resizing, but magick (IM 7) does. So if you do not want the masking, use IM 6 convert.

**Reply by jetpack1116** [reply]
@fmw42 
Ok.

When I test this I use IM Ver 7.1.0-2.

I will try this in IM 6.

**Comment by dlemstra** [comment]
Adding the image in parts for future reference:

[testPSDimage.001.zip](https://github.com/ImageMagick/ImageMagick/files/7192680/testPSDimage.001.zip)
[testPSDimage.002.zip](https://github.com/ImageMagick/ImageMagick/files/7192682/testPSDimage.002.zip)
[testPSDimage.003.zip](https://github.com/ImageMagick/ImageMagick/files/7192686/testPSDimage.003.zip)
[testPSDimage.004.zip](https://github.com/ImageMagick/ImageMagick/files/7192688/testPSDimage.004.zip)
[testPSDimage.005.zip](https://github.com/ImageMagick/ImageMagick/files/7192690/testPSDimage.005.zip)
[testPSDimage.006.zip](https://github.com/ImageMagick/ImageMagick/files/7192691/testPSDimage.006.zip)

**Comment by fmw42** [comment, marked as answer]
Here is what is happening. The first alpha channel in the PSD file will be read by ImageMagick (when no background transparency). This is confirmed by `magick identify -verbose testPSDimage.psd`.

When immediately saving as JPG, the alpha channel is turned off, because JPG does not support transparency. So the saved JPG will not show the masking from the alpha channel as in the following command:

`magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb result1.jpg`

however is saved as PNG, one will get transparency from the alpha channel as in the following command:

`magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb result1.png`

If one desires to have the masking as black in the saved JPG, then add -background black -flatten to the command before saving:

`magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -background black -flatten result2.jpg`

Now when one does some operation such as -resize or -blur, etc before saving to JPG, one does see the black from the alpha channel. This happens because the alpha channel is there and used, but before writing to JPG, the output is initialized to zero, then the resize image with alpha is overlaid and then the alpha channel is turned off for JPG. This is the case for the OP's original command:

`magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -resize 1920x1920 result3.jpg`

**So to get a resized JPG output without the black masking from the alpha channel, simply turn alpha off before the resize:**

`magick -define psd:alpha-unblend=off testPSDimage.psd -auto-orient -colorspace srgb -alpha off -resize 1920x1920 result4.jpg`

**Reply by jetpack1116** [reply]
@fmw42 
This is a very detailed explanation of the solution.
Thank you very much.
